src/sage/rings/finite_rings/hom_finite_field_givaro.pyx

```diff
@@ -105,7 +105,7 @@ cdef class SectionFiniteFieldHomomorphism_givaro(SectionFiniteFieldHomomorphism_
             sage: K.<T> = GF(3^4)
             sage: f = FiniteFieldHomomorphism_givaro(Hom(k, K))
             sage: g = f.section()
-            sage: g(f(t+1))
+            sage: g(f(t+1)) # random
```

Comment:
There are a great number of added `# random` tags, are those really random? Isn't this test checking an important identity?

Reply:
Yeah, I am uncomfortable with this, but all these doctests seem to depend on the old, undocumented behaviour of `.any_root()` being deterministic in certain places. Before, the end of the recursion would be a call of `f.roots()[0]`, which fully factors what remains and always returns the same root. 

Now, we recursively call `f.any_root()` and this uses CZ for the splitting and so is not deterministic. 

In my opinion "any_root()" should not be deterministic, but if we want these to be fixed then we need to have that the method in `hom` calls `f.roots()[0]` instead of the current `any_root()`.

I would love to have more advice on this though!

Reply:
A lot of these tests really are `# random` since field embeddings aren't unique: They were deterministic due to a particular implementation choice, but that wasn't really guaranteed anywhere and @GiacomoPope's new algorithm happens to be (1) better, and (2) randomized.
This particular test, however, seems like it should still be deterministic: Here `g` is supposed to be a left inverse of `f` by construction, and the test tests that by evaluating the composition on a particular input value.

Reply:
Sorry for my confusion above, I did not notice that I had over-zealously added random tags to these special tests. I have removed this in the latest commit.